### PR TITLE
Tuya TS004x + avoid false multiple press events + send ZCL default response for switches

### PR DESCRIPTION
**Pull request by sebres**
This PR is an accumulation of several tasks and solves following problems:
 
- implements device request of Tuya TS004x (or Zemismart ZigBee Switch) ~as well as fixes mistaken pairing of the devices as a light~ (not a part of this PR now at @Smanar request)  (#3611);
- avoids false multiple press events fired for (unanswered) events with the same frame sequence number (#3611, #4227);
- implements sending of ZCL default response for switch devices expecting that (#3611, #4227), so prevent repeated ZCL frame from device and a latency (up to 3 seconds) before device becomes able to notify about next click, if it does not receive an ack.

### CODE REVIEW

**Comment by Smanar** [comment]
@sebres you have not started from the master.
The code you have used from my branch must not be used here, he is for time synchronisation, not used for your device, and I need to rewrite it after an other PR was validated.

Will be better if you can make your PR from the official actual master, pls.

**Comment by sebres** [comment]
Sure... but without your branch it is useless, because I was unable to pair TS0044 as a switch (it was light), so I don't really know whether this PR would make sense.
But OK, I'll rebase my changes against master and push it here.

**Comment by sebres** [comment]
Done. And I saved original branch as `tuya-with-pairing` in my repo (for the record or other future tests).

**Comment by Smanar** [comment]
Nice thx, about the tuya branch you have tested, It don't impact direclty the device detection, you need to make 1 inclusion, the device is see as light but deconz save some information on the db, and on the second one, it can reconize the device and making switch device, it s a long story ....

But yes, it solve random tuya working mode, that can impact your device, but I m sorry, need to make 3 PR in the good order, so I need to wait for the n°1 was validated before making the 2 others.

**Comment by sebres** [comment]
> you need to make 1 inclusion, the device is see as light but deconz save some information on the db, and on the second one, it can reconize the device and making switch device, it s a long story

Nope. You said it already in #3611 and my answer was "No, it does not". I can't confirm described behavior: no matter how many times I paired it (via deCONZ or Phoscon Web-App), tried to remove, whatever, it remaining light (with resulting polling issue, etc) or even untyped, but it was never a switch.
Once I used plugin built from your tuya branch, it helps immediately.

> but I m sorry, need to make 3 PR in the good order

No problem, it can take as long as needed (and myself is switched to zigpy/ZHA, so I don't use deconz either way).

**Review by ebaauw — CHANGES_REQUESTED** [review]
This is a BAD idea, as it will break the auto-repair of bindings and attribute reporting, see https://github.com/dresden-elektronik/deconz-rest-plugin/issues/4572#issuecomment-806230246.

**Comment by sebres** [comment]
I can only repeat - "sorry but - No".
See https://github.com/dresden-elektronik/deconz-rest-plugin/issues/3611#issuecomment-782048480
Feel free to close the PR (FWIW in-between I completely switched to ZHA).

**Comment by manup** [comment]
I'm keeping this in the 2.11.x milestone for reference and will create a smaller PR to fix at least the *ZCL Default Response* issue.

> https://github.com/dresden-elektronik/deconz-rest-plugin/issues/3611#issuecomment-782048480

>* my engineers spirit refuses to implement such nursery-school methods, which may be fully legal in proof of concepts or some test branches, but do not belong in ready products at all.
>* and last but not least, I have absolutely no interest in possible association of me with an author of such dirty workarounds.

I understand and dislike the current hacky patchwork of device specific things which should be generic as well. This is one of the reasons why discussion in https://github.com/dresden-elektronik/deconz-rest-plugin-v2  and development in https://github.com/manup/deconz-rest-plugin/tree/device_descriptions is being made.

TLDR: The  future `Device` approach is meant to have minimal C++ code with knowledge about specific devices and is generic wherever possible and testable. All device specifics are taken from Device Description Files (DDF) in JSON Format.

Albeit quite some functionality is already working, we ain't there yet and it will take some time to get there.

That's why currently ugly device specific hacks are unavoidable, since little changes like in this PR can cause major trouble. Therefore we limit the "attack surface" to devices where we are certain it works for them without breaking otherwise tested and working devices in ten thousands of setups (which we did in the past with major headaches).

**Comment by sebres** [comment]
Sure. Just I'd like to see any reference to the spec where it is expressly allowed or rather serving any needs to consider a repeated incoming event (with the same sequence number)...

From other point of view you surely understand my intention to avoid such hacks, don't you? With other words it is allowed to me to reject a change call for my PR violating my own internal "policies".

> That's why currently ugly device specific hacks are unavoidable

Why?

somewhere in json for such devices:
```diff
+  "options": ["send_zcl_default_response": true]
```
somewhere in code:
```diff
     // default device flags enum:
+    sendZclDefaultResponse = 0x00010000,
...
     // flags of found device from json options:
+    if (...->json->get("send_zcl_default_response", false)) device->Flags |= deviceFlags.sendZclDefaultResponse;
...
+    if (device->Flags & deviceFlags.sendZclDefaultResponse) {
         if ( ((zclFrame.frameControl() & (deCONZ::ZclFCClusterCommand|deCONZ::ZclFCDisableDefaultResponse)) == deCONZ::ZclFCClusterCommand) &&
              !(ind.dstAddress().isNwkBroadcast()) )
         {
             sendZclDefaultResponse(ind, zclFrame, deCONZ::ZclSuccessStatus);
         }
+    }
```
That's it.
Just because I'm not really familiar with your code-base it is not so simple for the newbie like me to extend it properly, as well as to consider every internal handling which can either conflict with the change or miss it on some places (especially in this pretty "modular" code).
One could simply consider my PR as a working PoC (tested with this and other devices) and integrate it as some base in your code **properly**, instead of strange invitation to change this PR in the way violating every common sense.
And the longer your code would grow in this hackish way the complex it will become (and not for newbies only, believe me).

BTW. Regardless the different opinions and although it was quasi my own suggestion (to close this PR if no interest), it is really contra productive to make it in this way - it will just annoy the devs, spending own time invested in test or PoC and at least show other interested people a non-constructive and indifferent attitude.
